### PR TITLE
fix(ec2): `WindowsVersion` enum

### LINKED ISSUE #29736: ❗ NOTICE(ec2): new `WindowsVersions` enum values used in `WindowsImage()` gives: `Deployment failed: Error [ValidationError]: Unable to fetch parameters [/aws/service/ami-windows-latest/Windows_Server-<Specific-Image>-<YYYY.MM.DD>] from parameter store for this account.`
Please add your +1 👍 to let us know you have encountered this
---

### Overview:

Present in v2.135.0-v2.136.0

The only impact to existing applications is a deprecation notice. You can safely ignore this.

#29435 updated the windows versions enums. The new enums suffixed with dates (i.e. `WINDOWS_SERVER_2022_ENGLISH_FULL_BASE_2024_02_14`) are not found when used in functions that extend `GenericSSMParameterImage`. The working enums were mistakenly deprecated (i.e. `WINDOWS_SERVER_2022_ENGLISH_FULL_BASE`).

```
const ami = new ec2.WindowsImage(ec2.WindowsVersion.WINDOWS_SERVER_2022_ENGLISH_FULL_BASE_2024_02_14);

const instance = new ec2.Instance(this, "WindowsInstance", {
  vpc: new ec2.Vpc(this, "WindowsVPC"),
  instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
  machineImage: ami,
});
```
When deployed gives the following error.

### Workaround:

Continue to use the wrongly deprecated enum values that are not suffixed with dates.

### Solution:

We are working on reverting the original change.

### PR DESCRIPTION
**Pull request by nmussy**
### Issue # (if applicable)

Closes #29736

### Reason for this change

Oversight on my part when #29435 was created. I'd assumed that Windows versions and their datestamped counterparts (e.g. `Windows_Server-2022-English-Full-Base` and `Windows_Server-2022-English-Full-Base-2024.02.14`) could be retrieved in the same manner, and that the non-datestamped versions had just been removed since they were not being listed by `ec2:DescribeImages`.

The non datestamped versions are actually SSM parameter name suffixes used to retrieve the latest AMI image ID for a given region:

```sh
$ aws ssm get-parameters --names /aws/service/ami-windows-latest/Windows_Server-2022-English-Full-Base | jq '.Parameters[] | .Value'
"ami-03cd80cfebcbb4481"
$  aws ec2 describe-images --image-ids ami-03cd80cfebcbb4481 | jq '.Images[] | .Name'
"Windows_Server-2022-English-Full-Base-2024.03.13"
```

### Description of changes

* Split `WindowsVersion` into a second enum, `WindowsSpecificVersion`
  * Each `WindowsSpecificVersion` corresponds to a "datestampless" value in `WindowsVersion`, e.g. `WindowsSpecificVersion.WINDOWS_SERVER_2022_ENGLISH_FULL_BASE_2024_02_14` and `WindowsVersion.WINDOWS_SERVER_2022_ENGLISH_FULL_BASE`
* Removed `@deprecated` tags for `WindowsVersion`s 
* Added `MachineImage.specificWindows` static method to lookup `WindowsSpecificVersion`s
* Updated documentation, removed references to deprecated versions
* Updated enum values (see d852649b15a5793bce132771670bdda1b2223486, 7942755d1e0bf6e55da0897670ed2e8b4c29d02f, 1bce2689ed5cb36dfc7d94e9e2718dee0adc85e5 and d7c3a7cca599d780f42c7b1b9ea067ff05f2958b)

### Description of how you validated changes

I've added unit and integration tests to validate both `MachineImage.specificWindows` and `MachineImage.latestWindows`.

The `WindowsSpecificVersion`  values were compared to the SDK results of `ec2:DescribeImages` with the following params:
```json
{
	"Owners": ["amazon"],
	"Filters": [{ "Name": "name", "Values": ["Windows_Server*"] }]
}
```

The `WindowsVersion` values were compared to the SDK results of `ssm:GetParametersByPath` with the following params:
```json
{
	"Path": "/aws/service/ami-windows-latest",
}
```

The parameters that did not start with `/aws/service/ami-windows-latest/Windows_Server` were ignored. Some are Amazon Linux images:
* `amzn2-ami-hvm-2.0.*`
* `amzn2-x86_64-SQL_2019_*`

Others are either EC2LaunchV2 or NitroTPM Windows images, neither currently supported by the CDK:
* `EC2LaunchV2-Windows_Server-2016-English-*`
* `TPM-Windows_Server-2016-English-*`
* `TPM-Windows_Server-2019-English-*`
* `TPM-Windows_Server-2022-English-*`

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-ec2/test/example.images.lit.ts [hunk 1]
```diff
@@ -3,16 +3,31 @@ import * as ec2 from '../lib';
...
   storage: ec2.AmazonLinuxStorage.GENERAL_PURPOSE,
   cpuType: ec2.AmazonLinuxCpuType.X86_64,
+  kernel: ec2.AmazonLinux2Kernel.KERNEL_5_10,
+});
+
+const amznLinux2023 = ec2.MachineImage.latestAmazonLinux2023({
+  edition: ec2.AmazonLinuxEdition.STANDARD,
+  cpuType: ec2.AmazonLinuxCpuType.X86_64,
+  kernel: ec2.AmazonLinux2023Kernel.KERNEL_6_1,
 });
```
Comment (nmussy) [inline packages/aws-cdk-lib/aws-ec2/test/example.images.lit.ts:18]: This can be moved to another PR, just noticed the deprecated method usage when updating the Windows example below

File: packages/aws-cdk-lib/aws-ec2/test/example.images.lit.ts [hunk 2]
```diff
@@ -3,16 +3,31 @@ import * as ec2 from '../lib';
...
   virtualization: ec2.AmazonLinuxVirt.HVM,
   storage: ec2.AmazonLinuxStorage.GENERAL_PURPOSE,
   cpuType: ec2.AmazonLinuxCpuType.X86_64,
+  kernel: ec2.AmazonLinux2Kernel.KERNEL_5_10,
+});
+
+const amznLinux2023 = ec2.MachineImage.latestAmazonLinux2023({
+  edition: ec2.AmazonLinuxEdition.STANDARD,
+  cpuType: ec2.AmazonLinuxCpuType.X86_64,
+  kernel: ec2.AmazonLinux2023Kernel.KERNEL_6_1,
```
Comment (nmussy) [inline packages/aws-cdk-lib/aws-ec2/test/example.images.lit.ts:17]: This can be moved to another PR, just noticed the deprecated method usage when updating the Windows example below

File: allowed-breaking-changes.txt
Comment (nmussy) [inline allowed-breaking-changes.txt:1]: Not thrilled about adding a ton of noise to the breaking changes list, but I don't think it would safe to not list them 
all individually and use something like `removed:aws-cdk-lib.aws_ec2.WindowsVersion.*`

Comment (nmussy) [reply allowed-breaking-changes.txt:1]: Might not be required, I'm trying to have `WindowsVersion` become a union of two new enums instead, `WindowsLatest` and `WindowsSpecific`. Will see if JSII doesn't complain too much

Comment (nmussy) [reply allowed-breaking-changes.txt:1]: Aborting this idea, it would be even messier:

```ts
/**
 * @deprecated use {@link WindowsLatestVersion} or {@link WindowsSpecificVersion} instead
 */
export const WindowsVersion = { ...WindowsLatestVersion, ...WindowsSpecificVersion };
export type WindowsVersion = typeof WindowsVersion[keyof typeof WindowsVersion];
```

```sh
API elements with incompatible changes:
err  - ENUM aws-cdk-lib.aws_ec2.WindowsVersion: has been removed [removed:aws-cdk-lib.aws_ec2.WindowsVersion]
err  - METHOD aws-cdk-lib.aws_ec2.MachineImage.latestWindows: argument version, takes aws-cdk-lib.aws_ec2.WindowsLatestVersion (formerly aws-cdk-lib.aws_ec2.WindowsVersion): could not find type aws-cdk-lib.aws_ec2.WindowsVersion [incompatible-argument:aws-cdk-lib.aws_ec2.MachineImage.latestWindows]
err  - INITIALIZER aws-cdk-lib.aws_ec2.WindowsImage.<initializer>: argument version, takes aws-cdk-lib.aws_ec2.WindowsLatestVersion | aws-cdk-lib.aws_ec2.WindowsSpecificVersion (formerly aws-cdk-lib.aws_ec2.WindowsVersion): none of aws-cdk-lib.aws_ec2.WindowsVersion are assignable to aws-cdk-lib.aws_ec2.WindowsLatestVersion | aws-cdk-lib.aws_ec2.WindowsSpecificVersion, could not find type aws-cdk-lib.aws_ec2.WindowsVersion, could not find type aws-cdk-lib.aws_ec2.WindowsVersion [incompatible-argument:aws-cdk-lib.aws_ec2.WindowsImage.<initializer>]
```

Unsurprisingly, the enum appears as being deleted, even though it retains its values and typing. The bigger issue is that I would need to introduce another breaking change by changing the type of `MachineImage.latestWindows`. 

Removing the unusable enum values added to `WindowsVersion` is not really an issue for users, but requiring them to change their existing stacks would be a huge pain for basically no gain.

So we're back to the original plan of removing the newly added datestamped `WindowsVersion` values

File: packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts [hunk 3]
```diff
@@ -488,991 +376,2108 @@ export enum WindowsVersion {
...
   WINDOWS_SERVER_2022_TURKISH_FULL_BASE_2023_12_13 = 'Windows_Server-2022-Turkish-Full-Base-2023.12.13',
+  /** 2024.01.16 version of {@link WindowsVersion.WINDOWS_SERVER_2022_TURKISH_FULL_BASE} */
   WINDOWS_SERVER_2022_TURKISH_FULL_BASE_2024_01_16 = 'Windows_Server-2022-Turkish-Full-Base-2024.01.16',
+  /** 2024.02.14 version of {@link WindowsVersion.WINDOWS_SERVER_2022_TURKISH_FULL_BASE} */
   WINDOWS_SERVER_2022_TURKISH_FULL_BASE_2024_02_14 = 'Windows_Server-2022-Turkish-Full-Base-2024.02.14',
+  /** 2024.03.13 version of {@link WindowsVersion.WINDOWS_SERVER_2022_TURKISH_FULL_BASE} */
+  WINDOWS_SERVER_2022_TURKISH_FULL_BASE_2024_03_13 = 'Windows_Server-2022-Turkish-Full-Base-2024.03.13',
   /** 2022 - End */
-}
+}
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts:2483]: nit: newline

File: packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts [hunk 4]
```diff
@@ -488,991 +376,2108 @@ export enum WindowsVersion {
...
+  WINDOWS_SERVER_2022_RUSSIAN_FULL_BASE = 'Windows_Server-2022-Russian-Full-Base',
+  WINDOWS_SERVER_2022_SPANISH_FULL_BASE = 'Windows_Server-2022-Spanish-Full-Base',
+  WINDOWS_SERVER_2022_SWEDISH_FULL_BASE = 'Windows_Server-2022-Swedish-Full-Base',
+  WINDOWS_SERVER_2022_TURKISH_FULL_BASE = 'Windows_Server-2022-Turkish-Full-Base',
+  /** 2022 - End */
+}
+/**
+ * Specific, datestamped Windows version name.
+ */
+export enum WindowsSpecificVersion {
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts:625]: Will this break folks currently using a timestamped version?

Comment (nmussy) [reply packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts:625]: The datestamped versions never worked with `WindowsVersion`, since the constructor uses `ssm:GetParameter`, looking for the `/aws/service/ami-windows-latest/<version>` key. Only non-datestamped versions keys are listed, and their corresponding value points to the latest AMI version, as shown here:
```sh
$ aws ssm get-parameters --names /aws/service/ami-windows-latest/Windows_Server-2022-English-Full-Base | jq '.Parameters[] | .Value'
"ami-03cd80cfebcbb4481"
$  aws ec2 describe-images --image-ids ami-03cd80cfebcbb4481 | jq '.Images[] | .Name'
"Windows_Server-2022-English-Full-Base-2024.03.13"
```

Comment (msambol) [reply packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts:625]: Thanks for clarifying 👍🏼

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  ec2.WindowsVersion.WINDOWS_SERVER_2022_JAPANESE_FULL_SQL_2017_WEB,
+  ec2.WindowsVersion.WINDOWS_SERVER_2016_ENGLISH_FULL_BASE,
+  ec2.WindowsVersion.WINDOWS_SERVER_2019_ENGLISH_FULL_BASE,
+  ec2.WindowsVersion.WINDOWS_SERVER_2022_ENGLISH_CORE_ECS_OPTIMIZED,
+
+  ec2.WindowsSpecificVersion.WINDOWS_SERVER_2022_JAPANESE_FULL_SQL_2017_WEB_2023_12_13,
+  ec2.WindowsSpecificVersion.WINDOWS_SERVER_2016_ENGLISH_FULL_BASE_2023_11_15,
+  ec2.WindowsSpecificVersion.WINDOWS_SERVER_2019_ENGLISH_FULL_BASE_2023_12_13,
+  ec2.WindowsSpecificVersion.WINDOWS_SERVER_2022_ENGLISH_CORE_ECS_OPTIMIZED_2024_01_10,
+];
```
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts:58]: nice test of both 👍🏼

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    ...instanceVersions.map((version) => new InstanceTestCase(
+      app,
+      generateStackName('instance', version),
+      { machineImage: getMachineImage(version), env },
+    )),
+  ],
+  enableLookups: true,
+});
+
+function getMachineImage (version: Version) {
```
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts:78]: nit: remove space after function name

**Review by msambol — APPROVED** [review]
I like the direction you took this. This LGTM.

File: packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts (hunk 3 above)
Comment (nmussy) [reply packages/aws-cdk-lib/aws-ec2/lib/windows-versions.ts]: Fixed 👍

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts (hunk 6 above)
Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts]: Fixed 👍

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  return Object.values(ec2.WindowsVersion).includes(version as ec2.WindowsVersion) ?
+    ec2.MachineImage.latestWindows(version as ec2.WindowsVersion) :
+    ec2.MachineImage.specificWindows(version as ec2.WindowsSpecificVersion);
+}
+
+function generateStackName (prefix: string, version: string) {
```
[the remark this answers is not in the document]
Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.windows-machine-image.ts]: Fixed 👍

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
A couple of things here before I start looking at specifics:
1. I see the addition of deprecated versions. We shouldn't be adding anything that's deprecated.
2. What's the actual value add of all these hyper specific versions? If we are going to need to maintain a hyper specific and extremely long list like this, we need a pretty compelling reason for it.

**Comment by nmussy** [comment]
Main use case I see for this PR is giving users an easier access to stable versions so their instances don't get replaced on deploy when there's a new latest version. I don't think this is a good enough reason to continue down this path, reverting the previous change with #29737 should be enough 👍

**Comment by nmussy** [comment]
Good call. For now, I'm going to close this PR. I'll open another one to re-update `WindowsVersion` with the correct deprecated and missing values, and another one to implement dated versions in a more sensible way
